File: src/main/java/io/github/svegon/capi/util/GeometryUtil.java
package io.github.svegon.capi.util;

import com.google.common.collect.Lists;
import io.github.svegon.utils.collections.ListUtil;
import net.minecraft.util.math.*;
import net.minecraft.util.shape.VoxelShape;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.stream.Stream;

public final class GeometryUtil {
    private GeometryUtil() {
        throw new UnsupportedOperationException();
    }

    public static final Box UNBOUND_BOX = new Box(Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY,
            Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY);
    public static final Vec3d ORIGIN_CENTER_VEC = new Vec3d(0.5, 0.5, 0.5);

    public static BlockPos axisMinPos(BlockBox box) {
        return new BlockPos(box.getMinX(), box.getMinY(), box.getMinZ());
    }

    public static BlockPos axisMaxPos(BlockBox box) {
        return new BlockPos(box.getMaxX(), box.getMaxY(), box.getMaxZ());
    }

    public static List<BlockPos> allInBox(BlockPos start, BlockPos end) {
        final int minX = Math.min(start.getX(), end.getX());
        final int minY = Math.min(start.getY(), end.getY());
        final int minZ = Math.min(start.getZ(), end.getZ());
        final int xLength = Math.max(start.getX(), end.getX()) - minX;
        final int yLength = Math.max(start.getY(), end.getY()) - minY;
        final int zLength = Math.max(start.getZ(), end.getZ()) - minZ;

        return ListUtil.iterate((i, j, k) -> new BlockPos(minX + i, minY + j, minZ + k),
                xLength, yLength, zLength);
    }

    public static List<BlockPos> allInBox(BlockBox box) {
        return allInBox(axisMinPos(box), axisMaxPos(box));
    }

    public static List<BlockPos> allInBox(BlockPos center, int range) {
        return allInBox(center.add(-range, -range, -range), center.add(range, range, range));
    }

    public static Stream<BlockPos> allInBoxStream(BlockPos from, BlockPos to) {
        return allInBox(from, to).stream();
    }

    public static Stream<BlockPos> allInBoxStream(BlockPos center, int range) {
        return allInBox(center, range).stream();
    }

    public static Stream<BlockPos> allInBoxParallelStream(BlockPos from, BlockPos to) {
        return allInBoxStream(from, to).parallel();
    }

    public static Stream<BlockPos> allInBoxParallelStream(BlockPos center, int range) {
        return allInBoxStream(center, range).parallel();
    }

    public static Stream<BlockPos> stream(BlockBox box) {
        return allInBoxStream(new BlockPos(box.getMinX(), box.getMinY(), box.getMinZ()),
                new BlockPos(box.getMaxX(), box.getMaxY(), box.getMaxZ()));
    }

    public static Stream<BlockPos> parallelStream(BlockBox box) {
        return stream(box).parallel();
    }

    public static List<BlockPos> neighboringBlocks(BlockPos pos) {
        return Lists.newArrayList(pos.up(), pos.down(), pos.north(), pos.east(), pos.south(), pos.west());
    }

    public static String toSimpleString(@NotNull final BlockPos pos) {
        return pos.getX() + " " + pos.getY() + " " + pos.getZ();
    }

    public static Vec3d start(Box box) {
        return new Vec3d(box.minX, box.minY, box.minZ);
    }

    public static Vec3d dimensions(Box box) {
        return new Vec3d(box.getLengthX(), box.getLengthY(), box.getLengthZ());
    }

    public static Box scale(final Box box, final double scale) {
        return new Box(Vec3d.ZERO, dimensions(box).multiply(scale)).offset(start(box));
    }

    public static Vec3d closestPoint(Vec3d eyes, BlockPos blockPos, VoxelShape blockShape) {
        return closestPoint(eyes, Vec3d.of(blockPos), blockShape);
    }

    public static Vec3d closestPoint(Vec3d eyes, Vec3d pos, VoxelShape blockShape) {
        return blockShape.getClosestPointTo(eyes.subtract(pos)).orElse(ORIGIN_CENTER_VEC).add(pos);
    }

    public static boolean rectContains(int left, int top, int width, int height, int x, int y) {
        if ((width | height) < 0) {
            // At least one of the dimensions is negative...
            return false;
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false;
        }
        width += left;
        height += top;
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y));
    }

    public static boolean rectContains(float left, float top, float width, float height, float x, float y) {
        if ((Float.floatToRawIntBits(width) | Float.floatToRawIntBits(height)) < 0) {
            // At least one of the dimensions is negative...
            return false;
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false;
        }
        width += left;
        height += top;
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y));
    }

    public static boolean rectContains(double left, double top, double width, double height, double x, double y) {
        if ((Double.doubleToRawLongBits(width) | Double.doubleToRawLongBits(height)) < 0) {
            // At least one of the dimensions is negative...
            return false;
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false;
        }
        width += left;
        height += top;
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y));
    }

    public static BlockPos toBlockPos(double x, double y, double z) {
        return new BlockPos(MathHelper.floor(x), MathHelper.floor(y), MathHelper.floor(z));
    }

    public static BlockPos toBlockPos(Vec3d pos) {
        return toBlockPos(pos.x, pos.y, pos.z);
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/util/GeometryUtil.kt
package io.github.svegon.mclientapi.util

import com.google.common.collect.Lists
import net.minecraft.core.BlockBox
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.util.Mth
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import net.minecraft.world.phys.shapes.VoxelShape
import org.joml.Vector2f
import java.util.stream.Stream
import kotlin.math.sqrt
import kotlin.streams.asStream

object GeometryUtil {
    val UNBOUND_BOX: AABB = AABB(
        Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY,
        Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY
    )
    val ORIGIN_CENTER_VEC: Vec3 = Vec3(0.5, 0.5, 0.5)

    fun centeredBox(center: BlockPos, range: Int): BlockBox {
        return BlockBox(center.offset(-range, -range, -range),
            center.offset(range, range, range))
    }

    fun stream(box: BlockBox): Stream<BlockPos> {
        return box.asSequence().asStream();
    }

    fun parallelStream(AABB: BlockBox): Stream<BlockPos> {
        return stream(AABB).parallel()
    }

    fun neighboringBlocks(pos: BlockPos): List<BlockPos> {
        return Lists.newArrayList(pos.above(), pos.below(), pos.north(), pos.east(), pos.south(), pos.west())
    }

    fun toSimpleString(pos: BlockPos): String {
        return pos.x.toString() + " " + pos.y + " " + pos.z
    }

    fun start(box: AABB): Vec3 {
        return Vec3(box.minX, box.minY, box.minZ)
    }

    fun dimensions(box: AABB): Vec3 {
        return Vec3(box.xsize, box.ysize, box.zsize)
    }

    fun scale(box: AABB, scale: Double): AABB {
        return AABB(Vec3.ZERO, dimensions(box).scale(scale)).move(start(box))
    }

    fun closestPoint(eyes: Vec3, blockPos: BlockPos, blockShape: VoxelShape): Vec3 {
        return closestPoint(eyes, Vec3.atLowerCornerOf(blockPos), blockShape)
    }

    fun closestPoint(eyes: Vec3, pos: Vec3, blockShape: VoxelShape): Vec3 {
        return blockShape.closestPointTo(eyes.subtract(pos)).orElse(ORIGIN_CENTER_VEC).add(pos)
    }

    fun rectContains(left: Int, top: Int, width: Int, height: Int, x: Int, y: Int): Boolean {
        var width = width
        var height = height
        if ((width or height) < 0) {
            // At least one of the dimensions is negative...
            return false
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false
        }
        width += left
        height += top
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y))
    }

    fun rectContains(left: Float, top: Float, width: Float, height: Float, x: Float, y: Float): Boolean {
        var width = width
        var height = height
        if ((java.lang.Float.floatToRawIntBits(width) or java.lang.Float.floatToRawIntBits(height)) < 0) {
            // At least one of the dimensions is negative...
            return false
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false
        }
        width += left
        height += top
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y))
    }

    fun rectContains(left: Double, top: Double, width: Double, height: Double, x: Double, y: Double): Boolean {
        var width = width
        var height = height
        if ((java.lang.Double.doubleToRawLongBits(width) or java.lang.Double.doubleToRawLongBits(height)) < 0) {
            // At least one of the dimensions is negative...
            return false
        }
        // Note: if either dimension is zero, tests below must return false...
        if (x < left || y < top) {
            return false
        }
        width += left
        height += top
        //    overflow || intersect
        return ((width < left || width > x) &&
                (height < top || height > y))
    }

    fun toBlockPos(x: Double, y: Double, z: Double): BlockPos {
        return BlockPos(Math.floor(x).toInt(), Math.floor(y).toInt(), Math.floor(z).toInt())
    }

    fun toBlockPos(pos: Vec3): BlockPos {
        return toBlockPos(pos.x, pos.y, pos.z)
    }
    fun getFacing(pos: Vec3): Direction {
        return Direction.getApproximateNearest(pos.x(), pos.y(), pos.z())
    }

    /**
     * @see net.minecraft.entity.Entity
     *
     *
     * @param eyes position of the eyes looking at the angle
     * @param target the target point where they are pointed
     * @return a new Vec2f(yaw, pitch) of the angles the eyes are rotated to look at the target
     */
    fun lookVector(eyes: Vec3, target: Vec3): Vector2f {
        return lookVector(target.subtract(eyes))
    }

    fun lookVector(target: Vec3): Vector2f {
        val xDiff = target.x
        val yDiff = target.y
        val zDiff = target.z
        val horizontal = sqrt(xDiff * xDiff + zDiff * zDiff)

        val pitch = Mth.wrapDegrees((Mth.atan2(yDiff, horizontal) * -57.2957763671875).toFloat())
        val yaw = Mth.wrapDegrees((Mth.atan2(zDiff, xDiff) * 57.2957763671875).toFloat() - 90.0f)

        return Vector2f(pitch, yaw)
    }
}
